### PR TITLE
[multistage] Add Worker SPI to Make Worker Assignment Configurable

### PR DESCRIPTION
**Pull request by ankitsultana**
This adds a simple Worker SPI to make the worker assignment configurable per-query.

The worker assignment strategy may be picked by using queryOptions. This PR only adds the default implementation. We'll soon be adding a worker assignment strategy for colocated joins.

The interface is marked as Evolving since we'll be working heavily on worker assignment over the next month or so.

**Test-Plan:** Existing UTs have quite a lot of coverage. I also tested this by running the quickstart on local.

### CODE REVIEW

File: pinot-query-planner/src/main/java/org/apache/pinot/query/planner/logical/worker/WorkerAssignmentStrategy.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.query.planner.logical.worker;
+
+import org.apache.pinot.query.planner.QueryPlan;
+
+
+public interface WorkerAssignmentStrategy {
+  void assignWorkers(QueryPlan queryPlan);
```
Comment (walterddr) [inline pinot-query-planner/src/main/java/org/apache/pinot/query/planner/logical/worker/WorkerAssignmentStrategy.java:25]:
```suggestion
  void assignWorkers(QueryPlan queryPlan, WorkerManager workerManager);
```

File: pinot-query-planner/src/main/java/org/apache/pinot/query/planner/logical/worker/WorkerAssignmentStrategyFactory.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+public class WorkerAssignmentStrategyFactory {
+  private final WorkerManager _workerManager;
+
+  public WorkerAssignmentStrategyFactory(WorkerManager workerManager) {
+    _workerManager = workerManager;
+  }
+
+  public WorkerAssignmentStrategy create(Map<String, String> queryOptions, long requestId) {
+    return new DefaultWorkerAssignmentStrategy(_workerManager, requestId);
+  }
```
Comment (walterddr) [inline pinot-query-planner/src/main/java/org/apache/pinot/query/planner/logical/worker/WorkerAssignmentStrategyFactory.java:34]: in order for the SPI to work. the constructor of all implementation of WorkerAssignmentStrategy should take the exact same argument for constructors. 

see. `QuerySchedulerFactor` it has 
```
private static QueryScheduler getQuerySchedulerByClassName(String className, PinotConfiguration schedulerConfig,
      QueryExecutor queryExecutor, ServerMetrics serverMetrics, LongAccumulator latestQueryTime) {
        // ...
    return (QueryScheduler) constructor.newInstance(schedulerConfig, queryExecutor, serverMetrics, latestQueryTime);
  }
```

Comment (ankitsultana) [reply pinot-query-planner/src/main/java/org/apache/pinot/query/planner/logical/worker/WorkerAssignmentStrategyFactory.java]: I haven't added the reflection logic yet. Was thinking of adding it in the PR where I add the colocated join strategy. But yes we should implement the SPI in such a way that the dependencies for all implementations should be the same.

On a side-note, do we want to be using reflection in the query hot-path?

[the remark this answers is not in the document]
Comment (ankitsultana) [reply pinot-query-planner/src/main/java/org/apache/pinot/query/planner/logical/worker/WorkerAssignmentStrategyFactory.java]: I made two improvements in the latest revision:

1. Made `WorkerAssignmentStrategyFactory::create` static and made the class static as well by making the ctor private. This is similar to `QuerySchedulerFactory`. I feel it's neat this way since it doesn't make sense to have to pass around both the WorkerAssignmentFactory and WorkerManager across classes. UT may become an issue later but we can deal with it once it becomes an issue.
2. I have kept `WorkerManager` as a required dependency for creating a WorkerAssignmentStrategy and the `requestId` is passed to the method call `assignWorkers` instead of being a member var of the strategy.

File: pinot-query-planner/src/main/java/org/apache/pinot/query/planner/logical/worker/WorkerAssignmentStrategy.java [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+import org.apache.pinot.spi.annotations.InterfaceStability;
+
+
+/**
+ * WorkerAssignmentStrategy selects the servers to be used for each stage of the query.
+ */
+@InterfaceStability.Evolving
+public interface WorkerAssignmentStrategy {
+  void assignWorkers(QueryPlan queryPlan, long requestId);
+}
```
Comment (walterddr) [inline pinot-query-planner/src/main/java/org/apache/pinot/query/planner/logical/worker/WorkerAssignmentStrategy.java:31]: can be address in separate PR but I think we should separate out the WorkerAssignment as a class object --> we shouldn't nested this information inside QueryPlan anymore.

**Review by walterddr — COMMENTED** [review]
lgtm overall. @61yao could you please take a look as well?

File: pinot-query-planner/src/main/java/org/apache/pinot/query/planner/logical/worker/WorkerAssignmentStrategy.java (hunk 3 above)
Comment (ankitsultana) [reply pinot-query-planner/src/main/java/org/apache/pinot/query/planner/logical/worker/WorkerAssignmentStrategy.java:31]: Yeah I am not sure what the correct layout is but this definitely feels a bit off. Later we may also choose to instead return a worker assignment instead of modifying the plan in-place

**Comment by ankitsultana** [comment]
Let's follow-up on this offline as we have planned @walterddr
